### PR TITLE
PLT-279: Make compileVarFresh take an annotation

### PR DESCRIPTION
**Pull request by zliu41**
`compileVarFresh` should not assume that the `VarDecl` it creates should be annotated `AnnOther`. The caller should decide that.

This PR also makes sure all calls to `TermBind` constructor uses the correct annotation, since that's what the inliner looks at. This doesn't affect any existing test because we aren't inlining many things, but it's the right thing to do.

### CODE REVIEW

File: plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs [hunk 1]
```diff
@@ -453,18 +453,23 @@ hoistExpr var t = do
...
                 (PIR.Def var' (PIR.mkVar AnnOther var', PIR.Strict))
                 mempty
 
             t' <- maybeProfileRhs var' =<< compileExpr t
-            -- See Note [Non-strict let-bindings]
+                -- See Note [Non-strict let-bindings]
             let strict = PIR.isPure (const PIR.NonStrict) t'
-
-            PIR.modifyTermDef lexName (const $ PIR.Def var' (t', if strict then PIR.Strict else PIR.NonStrict))
+                -- Use `t'`'s annotation on `varAnnotated`, which ensures that if `t'` is marked
```
Comment (effectfully) [inline plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs:466]:
```
`t'`'
```

DOMINATE

**Review by effectfully — APPROVED** [review]
It's kinda unfortunate we have this duplication of annotations. Almost feels like we should always look at the term and not the binding annotation.

**Comment by effectfully** [comment]
(Don't take my approval too seriously though, I'm not too familiar with the inliner)

**Review by michaelpj — COMMENTED** [review]
I continue to be unconvinced by the argument that we need to do this annotation propagation at all. We don't mark any terms randomly as `AnnInline`, we only use it in the builtins support where we directly put the right annotation on the variable declaration. So why do we need to do this?

**Comment by zliu41** [comment]
@michaelpj Right now, `AnnInline` literally only works on builtins, and nothing else. In other words marking anything else `AnnInline` has no effect.

In my view, any time you mark any Term with `AnnInline`, and then if that term appears in the RHS of a term-let, then that term-let should be inlined. Or, if `AnnInline` is indeed only intended for `Builtins`, then we should rename it or add a note to make it clear. What do you think?

**Comment by michaelpj** [comment]
> In my view, any time you mark any Term with AnnInline, and then if that term appears in the RHS of a term-let, then that term-let should be inlined. Or, if AnnInline is indeed only intended for builtins, then we should rename it or add a note to make it clear. What do you think?

Isn't this the same discussion we already had? I argued for the annotation being meaningful on the _variable declaration_, you argued for having it be meaningful on the term. I think we agreed to say that it's the variable declaration annotation that's meaningful... so what are we doing here? Unless we're inconsistent and we were to start assuming that the annotation on the term was meaningful (which we currently don't, AFAICT), then there's no reason to do this, and it just _encourages_ us to be inconsistent. At the moment there's a simple rule: if you want it to inline, make sure you annotate the variable declaration. That's it!

**Comment by zliu41** [comment]
> Isn't this the same discussion we already had?

What we discussed is for builtins. In `defineBuiltinTerm`, we have the `var` and the `term`; we can annotate either one, and we decided to annotate the `var`.

But in the general case in `compileExpr`, it's a different story. When you compile a `GHC.Let`, you first compile the rhs. When compiling the rhs, the compiler may, for some reason, decide that rhs should be inlined (e.g., it is "const" or whatever). The compiler of the rhs does not have access to the `VarDecl`, so it can only mark the compiled rhs `AnnInline`, and then `compileExpr` should create the `VarDecl` based on that annotation.

> make sure you annotate the variable declaration.

That's exactly what I'm trying to do, but like I said above, in `compileExpr` we don't know what to annotate the `VarDecl`, unless we look at the compiled rhs.

**Comment by zliu41** [comment]
Since it is unclear whether this will ever be useful, I'm happy to just update the Note to clarify this for now. This is no hill to die on.

**Comment by michaelpj** [comment]
> But in the general case in compileExpr, it's a different story. When you compile a GHC.Let, you first compile the rhs. When compiling the rhs, the compiler may, for some reason, decide that rhs should be inlined (e.g., it is "const" or whatever). The compiler of the rhs does not have access to the VarDecl, so it can only mark the compiled rhs AnnInline, and then compileExpr should create the VarDecl based on that annotation.

But this is pretty hypothetical. I can equally argue that decisions about inlining will always be made a the binding site, and so we'll never need this! Indeed, I think what you are describing is pretty odd, actually: `compileExpr` doesn't know if it's compiling an expression which is the "top level" of a binding, so it has no idea whether marking it inline would even do anything.

**Comment by zliu41** [comment]
Ok, I'll back out the changes to `compileExpr` and update the Note.

**Comment by zliu41** [comment]
@michaelpj Updated.

**Review by michaelpj — APPROVED** [review]

**Pull request merged**